torch/_inductor/quantization.py

```diff
@@ -0,0 +1,233 @@
+    gm.graph.lint()
+    gm.recompile()
+
+
+def quantization_pre_grad_pass(gm: torch.fx.GraphModule, example_inputs):
+    # skip if gm is not a quantized graph module
+    if not (_is_quantized_graph_module(gm) and _is_cpu(example_inputs)):
+        return gm
+    gm.graph.eliminate_dead_code()
+    gm.recompile()
+
+    # Fuse `quant_per_channel - weight` and replace the original fp32 weight with quantized one
+    _pre_quantize_weights(gm)
+
+    return gm
+
+
+aten = torch.ops.aten
+prims = torch.ops.prims
+quantized_decomposed = torch.ops.quantized_decomposed
+dequantize_per_channel = quantized_decomposed.dequantize_per_channel.default
+
+"""
+dequantize activation:
+    x = x.to(fp32)
+    x = x - zero_point
+    x = x * scale
+"""
+dequantize_activation_pattern = CallFunction(
+    aten.mul.Tensor,
+    CallFunction(
+        aten.sub.Tensor,
+        CallFunction(
+            prims.convert_element_type.default, Arg(), Arg()  # args = x, torch.float32
+        ),
+        Arg(),  # zero point
+    ),
+    Arg(),  # scales
+)
+
+dequantize_weight_pattern = CallFunction(
+    dequantize_per_channel,
+    Arg(),  # weight
+    Arg(),  # scales
+    Arg(),  # zero point
+    Arg(),  # axis
+    Arg(),  # lower limit
+    Arg(),  # upper limit
+    Arg(),  # dtype=torch.int8
+)
+
+aten_conv_pattern = CallFunction(
+    aten.convolution.default,
+    dequantize_activation_pattern,
+    dequantize_weight_pattern,
+    Arg(),  # bias
+    Arg(),  # stride
+    Arg(),  # padding
+    Arg(),  # dilation
+    Arg(),  # transposed
+    Arg(),  # output_padding
+    Arg(),  # groups
+)
+
+"""
+quantize output:
+    scale = 1 / scale
+    scale = 1.0 * scale
+    output = round(output * scale)
+    output = output + zero_point
+    output = clamp_min(output, 0)
+    output = clamp_max(output, 127)
+    output = output.to(uint8)
+"""
+quantize_conv_output_pattern = CallFunction(
+    prims.convert_element_type.default,
+    CallFunction(
+        aten.clamp_max.default,
+        CallFunction(
+            aten.clamp_min.default,
+            CallFunction(
+                aten.add.Tensor,
+                CallFunction(
+                    aten.round.default,
+                    CallFunction(
+                        aten.mul.Tensor,
+                        aten_conv_pattern,  # output of conv
+                        CallFunction(
+                            aten.mul.Tensor,
+                            CallFunction(
+                                aten.reciprocal.default, Arg()  # arg = scales
+                            ),
+                            Arg(),  # 1.0
+                        ),
+                    ),
+                ),
+                Arg(),  # zero point
+            ),
+            Arg(),  # 0
+        ),
+        Arg(),  # 127
+    ),
+    Arg(),  # dtype=torch.uint8
+)
+
+
+@register_lowering_pattern(quantize_conv_output_pattern)
+def qconv(match: Match, *args, **kwargs):
+    """
+    There are 24 args. They are
```

Comment:
With this many args, it might be easier to use `KeywordArg("name")` so you don't need to remember the order they come in.

Reply:
Thanks. It's fixed.